### DISCUSSION TITLE
THEM WORDS vs PRONOUNs in Italian and Spanish Libs

### DISCUSSION THREAD

**Discussion by tajmone** [alan-if/alan-i18n#51] [New Locales]
The _[Predefined Player Words]_ of the **THEM WORDS** have always been a bit of a conundrum for me (see: alan-if/alan-docs#127).

First of all, I'm not sure if **THEM WORDS** can end up clashing with plural PRONOUNs, e.g. if I define the PRONOUN of a plural actors entity as THEM (e.g. "the goons" and "the cops"), how is the parser going to distinguish in a command like 'x them' whether 'THEM' is referring to a last referenced plural actor or to the last multiple parameters used (e.g. in a previous 'x the cops and the goons') — the 'x them' command could be referring to either just the "goons" (since it was the last object referred to, and it has PRONOUN 'THEM') or to both "cops" and "goons" (if we interpret THEM as referring to all the multi-parameters referred to).

I haven't carried out any tests in this respect, but @thoni56 might throw some light into the parsing mechanics, logic and priorities here.

The other problem pertains how **THEM WORDS** are problematic to handle in both Italian and Spanish, where these type of references usually come in the form of suffixes:

> &gt; x la mela E la pera ; (_eat the apple AND the pear_)
>
> &gt; mangiaLE ; (_eat THEM_) where 'LE' => THEM (f.p.)

In Spanish the problem is similar. So, without some means of intervening in the pre-parsing process, and being able to detect those suffixes and force a reparsing (`mangiale` &rarr; `mangia` + `them`) I'm not sure we can make use of **THEM WORDS** in the IT/ES libraries.

Inform Italian & Spanish are able to handle these suffix pronouns and split them, so IF players from these locales are used to being able to impart commands containing suffixed pronouns, which somehow plays against us, since acquired playing habits can be hard to drop.

## Spanish Foundation Status

Another problem worth mentioning is the current status of the `gramática.i` module in the Spanish library.

### Pronouns as `SYNONYMS`

Currently, we're defining all these pronouns as SYNONYMS of 'IT':

```alan
Synonyms
  eso, esos, esa, esas, ello,
  aquello, aquella, aquellos, aquellas, lo = it.
```

So if a player types 'x esos' the parser will try to match IT, instead of THEM.

It probably would make more sense to create two sets of SYNONYMS, one for singular pronouns, the other for the plural ones:

```alan
Synonyms eso, esa, ello, aquello, aquella, lo = IT.
Synonyms esos, esas, aquellos, aquellas = THEM.
```

which would probably guide the parser toward a better understanding of the player intention.

### Pronouns as `PRONOUN`s

But then I was wondering whether we should drop those SYNONYM and define proper PRONOUNs for each entity during initialization, so that each entity is given the correct PRONOUN according to GNA.

But if we do so, what happens to THEM words? Would we still need the above mentioned SYNONYMS? Would they conflict? (original doubts back in the scene)

@Rich15, are there valid usages of the detached pronouns in Spanish, which would justify their use as **THEM WORDS**?

Example:

> &gt; x agua Y café
>
> &gt; toma esos

Is that acceptable in Spanish?

In Italian it would sound _really_ bad, so much so that I doubt a player would even attempt that (instead of "prendiLI") unless he/she was instructed that it's the only way to make it work with the parser (they might just as well use THEM instead).

Are there any combinations where using the detached pronouns makes sense?

E.g. in Italian when dealing with verbs taking multiple objects you can end up having to use the detached forms:

> &gt; prendi la mela E il caffé
>
> &gt; dalLI a essi

But even if the above is acceptable, in Italian things get more complicated because in the above sentence we could have used "loro" instead of "essi". But "loro" also means "theirs" (e.g. "i loro soldi" &rarr; _their money_), so this is yet another example where creating a SYNONYM would prevent other legitimate uses ("loro" could be used to disambiguate between same named objects that differ in ownership: "their car", "my car").

Unfortunately the Italian grammar and lexicon is full of such ambiguities that are _very_ hard to cover in any IF language — which is probably why there are so few Italian IF libraries, even though the Italian IF community has been one of the most prolific ones, especially in the past were you could buy at the new-agent monthly magazines with audio-cassettes containing Italian text adventures to play on home computers like the C64, ZX-Spectrum, etc., all home-brew adventures, including their engines.

In that respect, the Spanish community has been more prolific, since they have produced more IF engines in Spanish than I've seen in any other language — there are dozens of _ad hoc_ Spanish IF engines, created only for Spanish (like Superglùs, to mention one), besides all the libraries translations from English to Spanish, something very impressive. As for the games, I have no idea how prolific it was.

Of course, many of these old engines and adventures belong to the InfoCom era, so they might not provide useful insights into advanced parsing techniques, but the modern engines like Inform do.

### Question About IT

If I've understood correctly, the word 'IT' is not hardcoded into the ALAN interpreter (unlike 'THEM' is, if you choose the English language).

Instead, it's hardcoded in the ALAN compiler, which will use it as a default fallback pronoun in case the author doesn't define one.

If the latter intuition is correct, it also means that since in the initialization module of the IT/ES libraries we're defining (or will be defining) a default PRONOUN on ENTITY (eso/esso, for ES/IT, respectively), because we need every entity to cover GNA, then there will be no trace of 'IT' in the final adventure story file...

If that's the case, then the following SYNONYMS (proposed above) will not be matching anything any longer:

```alan
Synonyms eso, esa, ello, aquello, aquella, lo = IT.
```

As you can see, when it comes to porting to other languages, these compiler defaults and predefined player words end up getting entangles with the library defaults that govern GNA and entities initialization.

So, right now I'm still not quite sure which of the above mentioned paths is the best one in terms of satisfying entities PRONOUNs initialization, on the one hand, and SYNONYMS for Pred.Player Words on the other (as well as other hard coded words).

-------------------------------------------------------------------------------

I've thrown a considerable load of coal on the fire in the thread ... Let's hope we can come up with some answers on how to handle these porting conundrums, because so far I haven't really managed to grasp which way to approach any of this.

But before v1.0.0 of the libraries is released, we'll have to decide on _some_ approach regarding how to implement pronouns and handle THEM WORDS.

[Predefined Player Words]: https://github.com/alan-if/alan-i18n/wiki/Predefined-Player-Words "See i18n Wiki article on 'Predefined Player Words'"

**Comment by Rich15** [comment]
### Pronouns

> It probably would make more sense to create two sets of SYNONYMS, one for singular pronouns, the other for the plural ones:
> ```
> Synonyms eso, esa, ello, aquello, aquella, lo = IT.
> Synonyms esos, esas, aquellos, aquellas = THEM.
> ```
>which would probably guide the parser toward a better understanding of the player intention.

Correct, it makes more sense.
> @Rich15, are there valid usages of the detached pronouns in Spanish, which would justify their use as THEM WORDS?
> 
> Example:
> ```
> > x agua Y café
>
> > toma esos
> ```
> Is that acceptable in Spanish?
> 
> In Italian it would sound really bad, so much so that I doubt a player would even attempt that (instead of "prendiLI") unless he/she was instructed that it's the only way to make it work with the parser (they might just as well use THEM instead).

It is acceptable, although a player would probably try "tómalos/tomarlos" instead (like in Italian).  "eso/esa/esos/esas" are all demostrative pronouns, so they would be assignated to unanimated objects.  In this case, a command like "toma eso/esos" would sound perfectly fine, but there should probably be an explicit instruction for the players.

> So, right now I'm still not quite sure which of the above mentioned paths is the best one in terms of satisfying entities PRONOUNs initialization, on the one hand, and SYNONYMS for Pred.Player Words on the other (as well as other hard coded words).

Since `PRONOUN`s could cause some trouble, maybe we should let "eso/esa..." etc., as Synonyms? Maybe define Pronouns for actors so you could do something like:
```
> x mujer

> habla con ELLA
```
But I have no clear idea on this. Which benefits would defining `PRONOUN`s bring vs letting Synonyms as they are?

The ideal scenario would be ALAN handling verb suffixes (E.g. `tómaLO`, `cómeLO`), but since this would require a lot of work, a temporary solution could be using demostrative pronouns.

**Reply by tajmone** [reply]
## Pronouns [ AMENDED! ]

> Correct, it makes more sense.

Then I'll start by applying this change, so at least IT maps to any singular object, and THEM to multiple parameters. In any case we're still in Beta, but it seems an improvement to me.

> Since PRONOUNs could cause some trouble, maybe we should let "eso/esa..." etc., as Synonyms? Maybe define Pronouns for actors so you could do something like:
> 
> &gt; x mujer

> &gt; habla con ELLA

That would already work with the change above, since all singular pronouns would be mapping to IT. The problem would really concern objects and actors which are plural in number, since the plural pronouns would be mapping to THEM, as in "multiple parameters last referenced", and not as entities which are plural in number (this affects not only plural actors, which are fairly rare, but also plural objects or objects with plural form like scissors, trousers, etc.)

Maybe the solution is simpler than I think, I might just test defining the pronouns in a dedicated branch and test what we get.

## Handling Suffix Pronouns via the Parser

> The ideal scenario would be ALAN handling verb suffixes (E.g. `tómaLO`, `cómeLO`), but since this would require a lot of work, a temporary solution could be using demostrative pronouns.

I don't think it has so much to do with hard work, it's more about not breaking the philosophy of ALAN being a simple language to use, for non-programmers. Solutions like those adopted in Inform are too "programmers like" for ALAN.

So it's more about coming up with an easy to use solution (i.e. for the users/authors). I've actually been giving this some though lately, and I _might_ have come up with an acceptable solution...

What we basically need is a fallback dictionary that the parser can look into when the parsed token doesn't match any syntax, direction or entity. It could then look at the suffixes map and try removing the indicated suffixes, check if what's left of the token initial part matches to a syntax, and if yes, then see what the dictionary wants the suffix to be reparsed as (i.e. non suffix pronoun).

Here's an example of how the dictionary would look like [[ **AMENDED:** 'IT' &rarr; 'THEM' ]] :

```
SUFFIX PRONOUNS
    los = eso. -- or 'THEM' (depending on the language).
    las = esa. -- or 'THEM' (depending on the language).
```

Then the parser encounters the player input `tomalos`:

1. The token `tomalos` doesn't match any direction or syntax.
2. Se if there's a "pronoun suffix" that matches the end:
    - YES: `los` is found at the end of `tomaLOS`, so:
        1. Split the token &rarr; `toma` + `los`
        2. Look up `los` in dictionary &rarr; `eso`
        3. Try parsing &rarr; `toma` + `eso`
        4. Matches a valid SYNTAX?
            - YES: Good to go.
            - NO: then it's malformed input.
    - NO: then it's malformed input.

The above syntax is not so hackish, it's just a dictionary. Moreover, this is something only library authors would use really, to add support for new languages, so it's not something the average author will have to bump into. Still, it's fairly elegant, and very similar to the syntax of SYNONYMS, except that in this case we're building a dictionary for pronouns suffixes substitutions (only used if parsing doesn't match).

@Rich15, does it seem a viable solution? Easy to use and, most important, that it would work for the goal?

What do you think about this @thoni56? Doesn't sound too different from the apostrophe elisions feature, and would most likely cover a good number of languages too.

**Reply by tajmone** [reply]
# Pronouns Tests Results

OK, I've done some testing and, as it turns out, having pronouns as SYNONYMS of THEM and also as definitions of the individual PRONOUN of entities _does not_ conflict in ALAN. Simply, ALAN will chose which to use based on the context.

Here's my test code:

-------------------------------------------------------------------------------

```alan
The park IsA location
  Name 'Central Park'.
End the park.

The flower IsA object at park.
  Is smellable.
  Pronoun it.
  Indefinite article "a"
End the.

The trees IsA object at park.
  Pronoun them.
  Indefinite form "some trees"
End the.

Syntax examine = examine (obj)*.

Syntax smell = smell (obj)*.

Add to every object
  Is not smellable.

  Verb examine
    Does "just $01."
  End verb.

  Verb smell
    Check obj is smellable
      Else "You can't smell $+1."
    Does
      "You smell $+1, nice fragrance."
  End verb.
End add to object.

Start at park.
```

-------------------------------------------------------------------------------

As you can see, the `trees` have PRONOUN THEM, while the `flower` has IT.

Here the resulting transcript:


```
Central Park
There is a flower and some trees here.

> ; Right now IT and THEM are not valid references (no past commands):
> examine it
I don't understand what you mean by 'it'.

> examine them
I don't understand what you mean by 'them'.

> ; But after a command involving an object with the pronoun it's OK:
> examine flower
Just a flower.

> smell it
You smell the flower, nice fragrance.

> ; Notice how here THEM refers only to the TREES:
> examine trees
Just some trees.

> smell them
You can't smell the trees.

> ; Now we use multiple parameters in a command:
> examine flower and trees
(flower) just a flower.

(trees) just some trees.

> ; So now THEM will refer to the last used multiple parameters:
> smell them
(flower) You smell the flower, nice fragrance.

(trees) You can't smell the trees.

> examine trees and flower
(trees) just some trees.

(flower) just a flower.

> smell them
(trees) You can't smell the trees.

(flower) You smell the flower, nice fragrance.

> ; But if we use a single parameter again...
> examine trees
Just some trees.

> ; then THEM refers to the TREES again, because last command was single param:
> smell them
You can't smell the trees.

```

So, PRONOUNs are only used to reference entities which have been the object (=parameter) of a previous command. But if the last command used multiple parameters, then ALAN will give precedence to THEM WORDS over PRONOUNs.

It's as simple as this, ALAN decides based on the context. It seems a natural choice, for that's most likely the player intention.

We only need to document this better, so it's clear.

I was thinking of jotting down some notes in a _Porting Guide_ document to include in the repository, explaining how to setup a grammar library module for non-English languages, with tips regarding _Predefined Player Words_, `Pronouns`, etc. — a bit like the Wiki, but using dynamic code examples.

So, it seems that the best solution is to give each entity it's correct pronoun during initialization, and _also_ define all plural pronouns as SYNONYMs of THEM. And let ARun decide which to use according to context.

Before doing that, I have one question @Rich15: In Italian some pronouns are meant to be used only with people (or sentient beings, like angles, talking animals, etc.) while others can be used on being and objects, and others only on objects (there's some disagreement on this rules, between "classic" and "modern" Italian, but I'd go for the broader interpretation where it suits us, and for the narrower where it prevents ambiguity).

Are the Spanish pronouns universal, or do they also follows similar rules?

**Reply by Rich15** [reply]
> That would already work with the change above, since all singular pronouns would be mapping to IT.

It would be different for actors, since the pronouns above are all demostrative, not personal.  So refering to a person as "esa/eso" would not sound good.  We could also define singular personal pronouns as Synonyms:
```
ella -> her
él -> him
```

> The problem would really concern objects and actors which are plural in number, since the plural pronouns would be mapping to THEM, as in "multiple parameters last referenced", and not as entities which are plural in number (this affects not only plural actors, which are fairly rare, but also plural objects or objects with plural form like scissors, trousers, etc.)
>
> Maybe the solution is simpler than I think, I might just test defining the pronouns in a dedicated branch and test what we get.

I see. Right now I have no idea how to handle it. Let me know if your tests are succesful and I'll let you know if I come up with something.

> What we basically need is a fallback dictionary that the parser can look into when the parsed token doesn't match any syntax, direction or entity. It could then look at the suffixes map and try removing the indicated suffixes, check if what's left of the token initial part matches to a syntax, and if yes, then see what the dictionary wants the suffix to be reparsed as (i.e. non suffix pronoun).
> 
> Here's an example of how the dictionary would look like:
> ```
> SUFFIX PRONOUNS
>     los = eso. -- or 'IT' (depending on the language).
>     las = esa. -- or 'IT' (depending on the language).
> ```
> Then the parser encounters the player input `tomalos`:
> 
> 1. The token `tomalos` doesn't match any direction or syntax.
> 2. Se if there's a "pronoun suffix" that matches the end:
>     - YES: los is found at the end of tomaLOS, so:
>       a. Split the token → toma + los
>       b. Look up los in dictionary → eso
>       c. Try parsing → toma + eso
>       d. Matches a valid SYNTAX?
>         - YES: Good to go.
>         - NO: then it's malformed input.
>     - NO: then it's malformed input.

That sounds great! I think this would work good for our purpose.

Just one little thing: "-los" and "-las" would be THEM, not IT.  Besides that, I think this is great idea.

**Reply by thoni56** [reply]
> ## Handling Suffix Pronouns via the Parser
> > The ideal scenario would be ALAN handling verb suffixes (E.g. `tómaLO`, `cómeLO`), but since this would require a lot of work, a temporary solution could be using demostrative pronouns.
> 
> I don't think it has so much to do with hard work, it's more about not breaking the philosophy of ALAN being a simple language to use, for non-programmers. Solutions like those adopted in Inform are too "programmers like" for ALAN.
> 
> So it's more about coming up with an easy to use solution (i.e. for the users/authors). I've actually been giving this some though lately, and I _might_ have come up with an acceptable solution...
> 
> What we basically need is a fallback dictionary that the parser can look into when the parsed token doesn't match any syntax, direction or entity. It could then look at the suffixes map and try removing the indicated suffixes, check if what's left of the token initial part matches to a syntax, and if yes, then see what the dictionary wants the suffix to be reparsed as (i.e. non suffix pronoun).
> 
> Here's an example of how the dictionary would look like [[ **AMENDED:** 'IT' → 'THEM' ]] :
> 
> ```
> SUFFIX PRONOUNS
>     los = eso. -- or 'THEM' (depending on the language).
>     las = esa. -- or 'THEM' (depending on the language).
> ```
> 
> Then the parser encounters the player input `tomalos`:
> 
> 1. The token `tomalos` doesn't match any direction or syntax.
> 2. Se if there's a "pronoun suffix" that matches the end:
>    
>    * YES: `los` is found at the end of `tomaLOS`, so:
>      
>      1. Split the token → `toma` + `los`
>      2. Look up `los` in dictionary → `eso`
>      3. Try parsing → `toma` + `eso`
>      4. Matches a valid SYNTAX?
>         
>         * YES: Good to go.
>         * NO: then it's malformed input.
>    * NO: then it's malformed input.
> 
> The above syntax is not so hackish, it's just a dictionary. Moreover, this is something only library authors would use really, to add support for new languages, so it's not something the average author will have to bump into. Still, it's fairly elegant, and very similar to the syntax of SYNONYMS, except that in this case we're building a dictionary for pronouns suffixes substitutions (only used if parsing doesn't match).
> 
> @Rich15, does it seem a viable solution? Easy to use and, most important, that it would work for the goal?
> 
> What do you think about this @thoni56? Doesn't sound too different from the apostrophe elisions feature, and would most likely cover a good number of languages too.

This is similar to what came to my mind when we discussed this earlier. So it's a viable solution, although I have not looked at the possibility to implement it yet.

One thing, is this only related to pronouns? In Swedish we have suffixes for definite articles, but that can be easily be solved with synonyms. But I was thinking maybe we could generalize this suffix replacement to apply to any kind of words. This way it would be kind of a synonym mechanism (being done before any parsing is attempted). Would that create problems for Spanish or Italian?

**Reply by tajmone** [reply]
> One thing, is this only related to pronouns? In Swedish we have suffixes for definite articles, but that can be easily be solved with synonyms. But I was thinking maybe we could generalize this suffix replacement to apply to any kind of words. This way it would be kind of a synonym mechanism (being done before any parsing is attempted). Would that create problems for Spanish or Italian?

The general idea is that these should be suffixes which are then translated to something else (no matter what really) before attempting reparsing, hence the idea of the map. That's because in Italian and Spanish if you were to just split the suffix from the word you'd obtain a definite article (in most cases) which would then be ignored by the parser (because they are also **NOISE WORDS**), so after splitting the suffix the parser must transform it in something else.

Probably, in most cases these pronouns suffixes would simply be reparsed as either IT or THEM, depending on their number, but since different languages have different grammars, the safest thing is to transform them in whatever the author wishes (and if the outcome is a SYNONYM, it will be handled as usual).

But I see where you're getting, i.e. that the choice of the KEYWORDS to handle this doesn't necessary need to be associated with pronouns, but maybe just suffixes.

**Reply by tajmone** [reply]
If I may add a couple of side-notes, I think that this prefix splitting functionality should have the following features:

1. Allow mapping the extracted suffix to any desired word (including an empty string), i.e. the split suffix should be reparsed normally, so if it's SYNONYM it's translated to it, if it's a NOISE WORD it's treated accordingly, etc., etc.
2. If after splitting it still fails to parse, attempt one more recursion, because often two pronouns are suffixed as a compound.

Practical examples:

The verbs "mangiatelo" (`mangia`+`te`+`lo`) and "mangialo" (`mangia`+`lo`) both mean _eat it_, but the latter is more like "_you_ eat it" (as in, instead of me), whereas the latter is just a plain direct order. The "te" in the former, is a reflexive form ("te"/"tu" &rarr; _you_ singular) which is often used in Italian, and sounds strange to non native speakers.

In this case, you'd want the parser to map it thus:

```
"te" => ""
"lo" => "IT"
``` 

so that the "te" is discarded — note that the "te" could be followed by "LI" (plural, THEM) or other pronouns, depending on the gender and number of the second verb object.

An example of multiple pronouns: "daglielo" (_give it to him/her_) which stands for `dai` + `IT` + `a lui`. In this case, beside the mapping of the extracted suffixes, there would also be a SYNONYM coming to rescue:

```
"ie" => "IT"
"lo" => "HIM"

SYNONYM dagl = dai. -- give
```

I wonder whether it wouldn't then just be simpler to allow splitting a prefix into multiple tokens directly, which would also avoid potential mismatches, since it ensures correct chaining of prefixes (since you basically end up with syllables, false positive could be easy):

```
"glielo" => "IT a HIM"    -- e.g. consegnaglielo (give it to him) 
"gliela" => "IT a HIM"    -- e.g. consegnagliela (give it to her) 
"glieli" => "THEM a HIM"  -- e.g. consegnaglieli (give them to him) 
""
``` 

etc. — there are more examples, but it was just to show that a single prefix could expand to multiple pronouns, and often even an implicit particle, as seen in the above examples, where the _TO_ is implicit in this type of form, yet the SYNTAX requires it to be implicit, which can't be achieved if each split prefix has to match to a single substitution word!
